### DISCUSSION TITLE
control the steering angle based on the output of a CNN

### DISCUSSION THREAD

**Discussion by noahzn** [bitcraze/discussions#561] [Q&A] [answered]
Hi, for example, if a model outputs 0, 1, or 2, which means turn left by -10 degrees, keep straight, and turn right by 10 degrees. How can I associate the output of a CNN with the control command of the crazyflie? Which file I should modify? Is there a similar script that I can take a look at? Thanks!

**Comment by hmllr** [comment, marked as answer]
Hi :)
I would do this using the application layer, similar as in this example: https://github.com/bitcraze/crazyflie-firmware/blob/master/examples/demos/app_push_demo/src/push.c
Just with the data from the AI-deck instead of from the multi-ranger (and for what you proposed you'd need to change the yaw command to absolute instead of velocity, however, in my experience a rate command for yaw is easier (you don't need to read your current yaw) and performs better if your CNN runs at a high enough rate, as yaw movement accumulates a lot of errors on the Crazyflie.
  
Then you also need to send over the data. The last time I did this we didn't have CPX yet, so we just sent the data over in custom UART packets. This would still be a possible solution for you. The other way (a bit less efficient, but easier to adapt and cleaner) would be to use the CPX protocol and send messages from the gap8 to an user application on the STM32. You find some infos about CPX here: https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/cpx/
and here: https://www.bitcraze.io/documentation/repository/aideck-gap8-examples/master/getting-started/cpx/

Hope this helps,
Hanna

**Reply by noahzn** [reply]
Hi! Thanks for your reply.
1. So, in the push.c, I think I can use `setpoint->mode.yaw = modeAbs; setpoint->attitude.yaw = yaw;`. Is this correct?
2. "a rate command for yaw is easier". Am I using the rate command in (1) now?

**Reply by hmllr** [reply]
1) yes, this is correct. However, keep in mind that this is an absolute yaw angle, so if you need to know your current angle to set it properly for your +/-10 degrees (which is not a problem, just some additional lines of code, you have examples on how to read log variables in the same push.c example).
2) yes, if you'd use the same code as in the push.c for the yaw steering you would command how fast you turn, not how much - which, depending on your network, might also be a valid option. Just something to think about, I don't know your application, maybe in your scenario that's not useful at all.

**Reply by noahzn** [reply]
Thank you very much for your quick reply! 
In my application I want to control the drone to avoid obstacles. I had a look at a control scheme using a depth map. So if there is an obstacle ahead on the left, the drone should steer a little to the right. The controlling scheme outputs `angular velocity` and in my test it's a number like `0.709`. (0 means go straight, positive numbers mean steering right, the larger the number, the larger the angle.)  Should I use the yaw command or a velocity command? I'm confused about them.

**Reply by hmllr** [reply]
What you usually would do is a combination - you want most likely want to slow down and turn away. There are different approaches in literature, e.g. PULP-DroNet which is also avoiding obstacles is scaling the velocity inversely to the collision probability and has a separate steering angle output. It might be possible to infer both from a steering angle though, as higher angles normally indicate higher collision probabilities (one of the newer DroNet papers might implement that, but I am not 100% sure). 
There are also approaches where you have discretized steps on how strong you turn (advertising my own work ;) https://www.bitcraze.io/2022/09/towards-autonomous-flight-on-nano-drones-with-multi-zone-depth-sensors/ however, this is not with the AI-deck).
I cannot recall any work with obstacle avoidance on nano-drones trying to turn by a specific angle right now, but that doesn't mean it doesn't exist.
With other words - there are lots of possibilities and you need to try out which works best for you :)

**Reply by noahzn** [reply]
Hi @hmllr , I'm now working on this. So, I need to call `cpxSendPacketBlocking`, for example, in the `classification.c`, and call `cpxReceivePacketBlocking` in the app layer `void appMain()`. Is this correct?

**Reply by knmcguire** [reply]
Pretty much! Don't forget to also initialize cpx as well. There is an example for this in the wifi streamer: https://github.com/bitcraze/aideck-gap8-examples/blob/master/examples/other/wifi-img-streamer/wifi-img-streamer.c

Since you already selected the answer, make a new discussion for any followup questiosn

**Reply by noahzn** [reply]
Thank you @knmcguire ! I will open a new discussion if I have further questions regarding this.
